Require disclosure of genAI usage in ai policy (#538)
astropy/astropy-project: policies/ai-policy.md

# Usage of Generative AI

Astropy maintainers welcome contributions, including those developed with the assistance of generative AI tools. However, the use of such tools does not change our expectations for contributor responsibility, engagement, or code quality.
The following apply to all contributions.

1. **Human ownership and accountability:**
    1. Contributors are responsible for all submitted content, regardless of whether generative AI tools were used.
    2. Contributors must be able to explain all changes during review.

>     3. Contributors are responsible for ensuring all submitted code complies with Astropy's licensing, regardless of whether generative AI tools were used.

Reviewer: ```suggestion
    3. **Licensing:** Contributors are responsible for ensuring all submitted code complies with Astropy's licensing, regardless of whether generative AI tools were used.
```

> 2. **Disclosure:** Committers must disclose whether they leveraged generative AI tools when making changes. This also helps downstream citations comply with academic conference policies that may require disclosure of genAI usage.

Author: added this line (the rest of the changes are bolding the topic sentences, which I can move to another PR if desired)

Reviewer 2: I think it is ok to do bold text here, but if this PR is somehow rejected, then the bold text would have to be a new PR. We shall see...

Author: ```suggestion
2. **Disclosure:** Contributors must disclose whether they leveraged generative AI tools when making changes (including autonomous agents). This also helps downstream citations comply with academic conference policies that may require disclosure of genAI usage.
```

> 3. **Authentic engagement:** The pull request process is collaborative and iterative. Contributors are expected to actively engage with reviewer feedback themselves. Copying and pasting replies to / from a generative AI tool does not count as engaging with the reviewer; contributors are expected to understand the changes they are proposing, not act as message courier for a third party. 

Author: Additional suggestion to cover autonomous genAI agents. Talking with a chatbot assuming it's a human is subtly draining and mentally taxing, as it can only ever approximate what an average developer might respond with, cannot handle abstraction, and sometimes will just lie and say it did something but the code changes don't actually reflect that. 

```suggestion
3. **Authentic engagement:** The pull request process is collaborative and iterative. Contributors are expected to actively engage with reviewer feedback themselves. Copying and pasting replies to / from a generative AI tool does not count as engaging with the reviewer; contributors are expected to understand the changes they are proposing, not act as message courier for a third party.

> [!IMPORTANT]
> Autonomous workflows ("agents") are not human contributors. Astropy reviewers expect to interact with the human controlling the account, or we will close the PR.
```

# Usage of Generative AI

Astropy maintainers welcome contributions, including those developed with the assistance of generative AI tools. However, the use of such tools does not change our expectations for contributor responsibility, engagement, or code quality.
The following apply to all contributions.

1. **Human ownership and accountability:**
    1. Contributors are responsible for all submitted content, regardless of whether generative AI tools were used.
    2. Contributors must be able to explain all changes during review.

>     3. **Licensing:** Contributors are responsible for ensuring all submitted code complies with Astropy's licensing, regardless of whether generative AI tools were used.

Reviewer 3: This isn't actually a change in this PR, but this line seems unenforceable to me. If I submit a PR with a new function and tests that were AI generated, how exactly do I "ensure that all the code complies with Astropy's licensing"?

Reviewer 4: I'm tempted to conclude that use of genAI is intrinsically incompatible with the ethics of FOSS, and probably copyright law too 🤔

Reviewer 5: On the latter, probs. 
On the former, https://lore.kernel.org/linux-media/CAHk-=wi4zC+Ze8e+p3tMv8TtG_80KzsZ1syL9anBtmEh5Z40vg@mail.gmail.com/

Reviewer 4: Linus isn't the Pope of FOSS, as far a I'm aware. And he seems widely disconnected with the _actual_ arguments against genAI, so I'm not feeling to bad that I disagree with him.

Reviewer 5: I'm not sure there's any Pope of FOSS, but continuing on this metaphor, there are definitely bishops, and archbishops given major FOSS projects. 
I don't think we should be disallowing the next set of demonstrably useful tools. 
We should make sure the work for us, and not us work for them.

Reviewer 4: Usefulness is not the point. The tech is an ecological disaster, and a powerful enabler to authoritarian regimes of the world. _These_ are the reasons why I want no part in it.

Reviewer 5: A perfectly fair personal position. I'm not conviced we should force contributors to make the same choice.

Reviewer 3: That post from Linus aligns perfectly with how I feel on almost every point. In particular, astropy is not a "social warrior" project and discussions related to the ethics of AI should take place in some other forum.

Author: What do you mean by "social warrior project" here? I'm also unclear what Linus meant by it. As far as I understand, Astropy, and open source in general, is based on ideals of community, collaboration, and the idea that software should be accessible and transparent, and that sounds a lot like a social stance to me. Personal convictions contribute to community sentiment, because persons make up the community; I don't think it's productive to dismiss that. In the end this software doesn't exist in a vacuum, it's made by and for people.

Putting aside the ethical and environmental issues however and focusing more on this PR, I still don't personally think entirely-LLM-generated pull requests are a net positive for any software product. LLM coding assistants, by their very nature, are not able to build abstract architectures without reflecting or mashing already-existing architecture from different projects with different goals. They can't plan ahead because they can't plan, or think in the first place. Without a human helming the change and PR, most of what I've seen from fully-autonomous PRs is band-aid solutions that don't account for future changes, and tailor-made modules that aren't maintainable, and all that turns into tech debt that a human has to deal with later. Not to mention, as I brought up in another comment, I've seen LLMs state with _utter confidence_ that they've taken care of some problem you bring up for them to fix, and then you check the actual code and they did nothing of the sort, or they added a test that passes green without actually covering the edge case it says it does, or the conditionals are so specific to fit your request that it fails to be useful for general cases. 

All of this checking and validation and not trusting the output is an additional mental load on reviewers, and I personally am of the conviction that purely-AI-generated pull requests are not worth it, which is why I added this note: https://github.com/astropy/astropy-project/pull/538#pullrequestreview-4743746382

Author: but again, I myself am not a regular reviewer of astropy, mostly just a downstream user. I don't mean to speak over anyone else's views.

Reviewer 4: I 100% adhere to @zacharyburnett's point. I'll just add that collaborative work is a social and political activity, especially in FOSS, and that politics cannot be waved away from it.

Reviewer 6: Howdy, I'm a maintainer over in JuliaAstro that happened to see this thread. I can only speak for myself, but as a person whose life is political by virtue of being a minority here, I just wanted to say that I deeply appreciate the thoughtful discourse being had by the astropy team

Reviewer 7: FWIW, I tend to agree with @zacharyburnett's perspective that everything we do has a social and community impact.  Astropy *is* a community project, and it is right and proper for us to consider the large-scale impacts of the things we do.

That said, to me this is an area where we are *very* far from community consensus.  @neutrinoceros has a very strongly-held view, that he has every right to have, and to advocate for that view in project discussions like this.  But I think this thread shows that plenty of other members of the community do *not* share that stance.  I for one, am one of them: I am very suspicious of both the questionable promises AI companies make, and have serious concerns of some of the knock-on effects (not the same as @neutrinoceros, FWIW, but different concerns to be discussed elsewhere)... but I also see very real benefit for *some* AI applications that seem like a net good.  Others in this and other discussion threads lie in other parts of that spectrum.  

So how I am going to choose to interpret @taldcroft's comment (but can be corrected if that's  mis-understanding!) is that it is not fair for one member (or even a handful) to have their views reflected in this policy if we have not reached community consensus.  

On this specific point, though, I tend to agree with @taldcroft that this is not enforceable, but I also think it should still be in here for now... but see my comment at the end about the pathway the CoCo wants to propose to address that.

> 2. **Disclosure:** If generative AI tools were used, contributors must describe their use in the change description. This also helps downstream citations comply with academic conference policies that may require disclosure of genAI usage.

Reviewer 5: Autocorrect on Macs / iPhone / Android /etc uses a transformer architecture trained on text corpus — making it a form of gen-ai. I think we need to be a little more careful about what we mean by "generative AI tools" since me typing this sentence on my computer and clicking tab to get the "ing" in "typing" used an on-device gen-ai. 

I'm not happy with my suggestion either, but maybe something like...

```suggestion
2. **Disclosure:** If generative AI tools were used for substantive portions of the contribution, contributors must describe their use in the change description. This also helps downstream citations comply with academic policies that may require disclosure of genAI usage.
```

Reviewer 8: fwiw, the orginal was better, no need for vague language of what classifies of "substantive".
I would urge you to keep the focus on the reason for this amandement to the policy. 
We don't need unsolicited PRs arguing and being dismissive of maintainers and reviewers. If you or TomA love coding with an agent, do it, you likely will do the review work, too of what comes out of it rather than pushing it raw on the other maintainers and be just a third party in the coversation. And then you could also declare the usage as you would expected to do for a paper and conference submission, too.

> 2. **Disclosure:** Contributors must disclose whether they leveraged generative AI tools when making changes (including autonomous agents). This also helps downstream citations comply with academic conference policies that may require disclosure of genAI usage.

Reviewer 5: We should add something, like a checkbox + fill in the blank to the Issue / PR templates!

Reviewer 2: A PR to https://github.com/astropy/astropy/blob/main/.github/PULL_REQUEST_TEMPLATE.md could accompany this PR, yes.

Author: from @taldcroft's suggestion above https://github.com/astropy/astropy-project/pull/538#issuecomment-5037168491, we can turn this from a simple yes / no question into a "describe how they were used":
```suggestion
2. **Disclosure:** If generative AI tools were used, contributors must describe their use in the change description. This also helps downstream citations comply with academic conference policies that may require disclosure of genAI usage.
```

> 2. **Disclosure:** Contributors must disclose whether they leveraged generative AI tools when making changes (including autonomous agents). This also helps downstream citations comply with academic conference policies that may require disclosure of genAI usage.

Reviewer 3: > This also helps downstream citations comply with academic conference policies that may require disclosure of genAI usage.

I don't have a clue what that means.

Author: if academic publishing requires disclosure of AI usage, I would think that also means people using astropy should be aware if a module or new version they're using has AI-generated code right? But maybe this isn't really the place to state that, I can remove it if desired:
```suggestion
2. **Disclosure:** Contributors must disclose whether they leveraged generative AI tools when making changes (including autonomous agents).
```

Reviewer 3: >  if academic publishing requires disclosure of AI usage, I would think that also means people using astropy should be aware if a module or new version they're using has AI-generated code right?

It is important to remember that AI-generated code is not a bright shiny line in the sand. Many developers use this to varying degrees, and there is already no way to say which subpackages in astropy have genAI code. I believe that in a matter of years, most astropy subpackages that are being actively maintained will have such code. So to my mind, we can already just make a blanket statement that astropy contains AI-generated code and folks can use that information as needed.

Reviewer 4: > I believe that in a matter of years, most astropy subpackages that are being actively maintained will have such code.

I'd rather that be a deliberate choice than a vague sentiment. Of course I'd personally prefer we take the _opposite_ choice.

> 3. **Authentic engagement:** The pull request process is collaborative and iterative. Contributors are expected to actively engage with reviewer feedback themselves. Copying and pasting replies to / from a generative AI tool does not count as engaging with the reviewer; contributors are expected to understand the changes they are proposing, not act as message courier for a third party.

Reviewer 5: I think this needs some polishing :). "not act as message courier for a third party." 😆

Reviewer 2: I think one of those LinkedIn profiles call themselves "agent sitters".

Reviewer 5: > "contributors are expected to understand the changes they are proposing"

is repetitive with "Contributors must be able to explain all changes during review." 

```suggestion
3. **Authentic engagement:** The pull request process is collaborative and iterative. Contributors are expected to actively engage with reviewer feedback themselves. Copying and pasting replies to / from a generative AI tool does not count as engaging with the reviewer.
```

The rest is already covered in text

Reviewer 5: It looks like we already had this in text, so it's fine to keep. But also, maybe good to polish.

Reviewer 8: Do keep this in, especially about the third party part.

Reviewer 7: I think based on the comment I'm about tot leave below re: CoCo discussions it's probably better to not make that change yet, but maybe in the follow-on PR (i.e., this is more about "philosophy" than policy, which we want to split up)
